### PR TITLE
[EE-IR] Support Local Functions

### PR DESCRIPTION
**Pull request by kandersen**
This commit introduces support for calling and referencing local functions and
objects in evaluate expression on the IR backend.

The primary incision is a lowering inserted after Local Declaration Lowering,
that uses the intermediate data structures recorded by LDL to rewrite calls to
local functions to the appropriate function in the binary, instead of predicting
the compilation strategy. The required changes to the rest of the pipeline
facilitate piping the required data around.

The key to this transformation is that _captures by the local function_ must be
introduced as _captures by the fragment function_, such that the evaluator
infrastructure can find the appropriate values at run-time. This is necessary
due to the strategy of compiling local functions to static functions instead of
closures.

Additional test coverage of stepping behavior support the corresponding changes
in the Evaluator, part of the Kotlin Debugger plug-in.

### CODE REVIEW

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt [hunk 1]
```diff
@@ -142,7 +142,13 @@ internal val localDeclarationsPhase = makeIrFilePhase(
                 private fun scopedVisibility(inInlineFunctionScope: Boolean): DescriptorVisibility =
                     if (inInlineFunctionScope) DescriptorVisibilities.PUBLIC else JavaDescriptorVisibilities.PACKAGE_VISIBILITY
             },
-            forceFieldsForInlineCaptures = true
+            forceFieldsForInlineCaptures = true,
+            postLocalDeclarationLoweringCallback = { (localFunctions, newParameterToOld, oldParameterToNew, newParameterToCaptured) ->
```
Comment (kandersen) [inline compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt:146]: @udalov I realize, preparing the PR, that the current version of this PR is order dependent: only the last invocation of LocalDeclarationsLowering's intermediate data structures will be recorded.

I think I just need to lift this to a map from containing IrFile/IrModule->Intermediate structure, and can quickly do so -- but do you agree with the overall approach of "state passing" via the context?

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt [hunk 2]
```diff
@@ -80,14 +83,15 @@ internal class ReflectiveAccessLowering(
...
+    // reflective access. We record these _before_ transformation, in order to
     // later predict the compilation strategy for fields. See the uses of
     // `fieldLocationAndReceiver`.
     val callsOnCompanionObjects: MutableMap<IrCall, IrClassSymbol> = mutableMapOf()
 
+    @OptIn(ObsoleteDescriptorBasedAPI::class)
     private fun recordCompanionObjectAsDispatchReceiver(expression: IrCall) {
-        if ((expression.dispatchReceiver as? IrGetObjectValue)?.symbol?.owner?.isCompanion == true) {
-            callsOnCompanionObjects[expression] = (expression.dispatchReceiver!! as IrGetObjectValue).symbol
+        if (((expression.dispatchReceiver as? IrGetField)?.symbol?.owner?.type as? IrSimpleTypeImpl)?.classifier?.descriptor?.isCompanionObject() == true) {
```
Comment (kandersen) [inline compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt:93]: Is there a better way to detect calls on companion objects? This feels extremely ad-hoc, but works.

File: compiler/testData/debug/localVariables/constructors/multipleConstructors.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
+
```
Comment (kandersen) [inline compiler/testData/debug/localVariables/constructors/multipleConstructors.kt:1]: As mentioned in the PR, these stepping tests are not directly related to this PR, but support the corresponding PR on the plug-in side by clarifying the stepping behavior of the JVM backend v/ the IR backend.

**Comment by kandersen** [comment]
Muted the new stepping tests on FIR -- spotted them as TeamCity failures :)

File: compiler/testData/debug/localVariables/constructors/multipleConstructors.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    constructor(p: Int): this() {
+        val a = 1
+    }
+
+    constructor(p1: Int, p2: Int): this()
+}
+
+fun box() {
+    Derived(1)
+    Derived(1, 1)
```
Comment (udalov) [inline compiler/testData/debug/localVariables/constructors/multipleConstructors.kt:16]: I know this is a very minor issue and not even related to the changes in the PR, but in such kind of tests I prefer to always use different values if possible, just to additionally check that we don't mix them up somehow. :) It also makes it a bit easier to read the expectations below. I.e. in this test there's `val a = 1`, so this could be
```
    Derived(2)
    Derived(3, 4)
```

File: compiler/testData/debug/stepping/kt15259.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
+// IGNORE_BACKEND: JVM_IR
+// IGNORE_BACKEND_FIR: JVM_IR
+// FILE: test.kt
+interface ObjectFace
+
+private fun makeFace() = object : ObjectFace {
+    //Breakpoint! from the Evaluate Expression test suite.
```
Comment (udalov) [inline compiler/testData/debug/stepping/kt15259.kt:7]: I'd replace these "Breakpoint!" lines with empty lines just to reduce confusion (one could think that this is somehow handled in compiler tests, which it's not)

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmBackendContext.kt [hunk 6]
```diff
@@ -47,6 +48,11 @@ class JvmBackendContext(
     val backendExtension: JvmBackendExtension,
     val irSerializer: JvmIrSerializer?,
 ) : CommonBackendContext {
+    var localFunctions: Map<IrFunction, LocalDeclarationsLowering.LocalFunctionContext> = mapOf()
+    var newParameterToOld: Map<IrValueParameter, IrValueParameter> = mapOf()
+    var oldParameterToNew: Map<IrValueParameter, IrValueParameter> = mapOf()
```
Comment (udalov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmBackendContext.kt:53]: Looks like this map is not used

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt [hunk 7]
```diff
@@ -386,23 +392,64 @@ private val jvmFilePhases = listOf(
...
+                buildLoweringsPhase(phases) then
+                generateMultifileFacadesPhase then
+                resolveInlineCallsPhase then
+                // should be last transformation
+                prepareForBytecodeInlining then
+                validateIrAfterLowering
+    )
+}
+
+private fun buildLoweringsPhase(
```
Comment (udalov) [inline compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt:425]: Sorry, I couldn't resist rewriting it a bit more in a functional style. It's not perfect though because requires explicit generic type arguments. :/
```
private fun buildLoweringsPhase(
    perModuleLowerings: List<Pair<String, List<NamedCompilerPhase<JvmBackendContext, IrFile>>>>,
): CompilerPhase<JvmBackendContext, IrModuleFragment, IrModuleFragment> =
    perModuleLowerings.map { (name, lowerings) -> performByIrFile(name, lower = lowerings) }
        .reduce<
                CompilerPhase<JvmBackendContext, IrModuleFragment, IrModuleFragment>,
                CompilerPhase<JvmBackendContext, IrModuleFragment, IrModuleFragment>
                > { result, phase -> result then phase }
```

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt (hunk 1 above)
Comment (udalov) [reply compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt:146]: It seems fine at first glance, but I would prefer if this was only done in case the fragment evaluation is enabled, and skipped if this is the normal compilation mode.

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt (hunk 2 above)
Comment (udalov) [reply compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt:93]: Seems fine, but let's avoid using descriptors, and computing same things multiple times:
```
    private fun recordCompanionObjectAsDispatchReceiver(expression: IrCall) {
        val dispatchReceiver = expression.dispatchReceiver as? IrGetField ?: return
        val dispatchReceiverType = dispatchReceiver.symbol.owner.type as? IrSimpleType ?: return
        val klass = dispatchReceiverType.classOrNull
        if (klass != null && klass.owner.isCompanion) {
            callsOnCompanionObjects[expression] = klass
        }
    }
```

File: compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/fragments/FragmentModuleGenerator.kt [hunk 8]
```diff
@@ -38,14 +38,28 @@ class FragmentModuleGenerator(
...
+                            generateSingleFile(DeclarationGenerator(it), ktFile, irModule)
+                        }
                     }
                 )
             }
         }
     }
 
+    private fun <T> generateInContextWithoutFragmentInfo(ktFile: KtFile, block: (GeneratorContext) -> T): T {
+        val symbolTableDecorator = (context.symbolTable as FragmentCompilerSymbolTableDecorator)
```
Comment (udalov) [inline compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/fragments/FragmentModuleGenerator.kt:51]:
```suggestion
        val symbolTableDecorator = context.symbolTable as FragmentCompilerSymbolTableDecorator
```

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FragmentLocalFunctionPatchLowering.kt [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+// whether the captures of the local function are a subset of the captures of
+// the fragment, and if not, introduces additional captures to the fragment
+// wrapper. The captures are then supplied to the fragment wrapper as
+// parameters supplied at evaluation time.
+internal class FragmentLocalFunctionPatchLowering(
+    val context: JvmBackendContext
+) : IrElementTransformerVoidWithContext(), FileLoweringPass {
+
+    override fun lower(irFile: IrFile) {
+        if (irFile.name != "fragment.kt") return
```
Comment (udalov) [inline compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FragmentLocalFunctionPatchLowering.kt:48]: Seems ad-hoc and kind of dangerous. :) Why is it needed?

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FragmentLocalFunctionPatchLowering.kt [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                            // in the call to the lowered function.
+                            val newParameter = declaration.addValueParameter {
+                                type = capturedValueSymbol.owner.type
+                                name = capturedValueSymbol.owner.name
+                            }
+
+                            context.state.recordNewFragmentCaptureParameter(
+                                newParameter.name.asString(),
+                                capturedValueSymbol.descriptor.type,
+                                capturedValueSymbol.descriptor
```
Comment (udalov) [inline compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FragmentLocalFunctionPatchLowering.kt:97]: Can we use IR-based descriptors here (and thus remove `@OptIn` on the function)?
```
                                capturedValueSymbol.owner.type.toIrBasedKotlinType(),
                                capturedValueSymbol.owner.toIrBasedDescriptor()
```

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt (hunk 2 above)
Comment (kandersen) [reply compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/ReflectiveAccess.kt:93]: Agreed! Done.

File: compiler/testData/debug/localVariables/constructors/multipleConstructors.kt (hunk 4 above)
Comment (kandersen) [reply compiler/testData/debug/localVariables/constructors/multipleConstructors.kt:16]: No, I agree! I did similar changes on the plug-in side previously because a big aggregate test made 10 assertions with the same constants :sweat_smile:

Comment (kandersen) [reply compiler/testData/debug/localVariables/constructors/multipleConstructors.kt:16]: Done :+1:

File: compiler/testData/debug/stepping/kt15259.kt (hunk 5 above)
Comment (kandersen) [reply compiler/testData/debug/stepping/kt15259.kt:7]: Done!

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmBackendContext.kt (hunk 6 above)
Comment (kandersen) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmBackendContext.kt:53]: Done.

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt (hunk 7 above)
Comment (kandersen) [reply compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt:425]: I can always appreciate a good fold! Done!

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FragmentLocalFunctionPatchLowering.kt (hunk 10 above)
Comment (kandersen) [reply compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FragmentLocalFunctionPatchLowering.kt:97]: Yes, that works!

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FragmentLocalFunctionPatchLowering.kt (hunk 9 above)
Comment (kandersen) [reply compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FragmentLocalFunctionPatchLowering.kt:48]: You're absolutely right, must have been an ad-hoc and kinda dangerous duct-tape solution to some intermediate state of the code :D 

I have no record of why I introduced it, and `visitSimpleFunction` right after is just as defensive.

Out it goes!

File: compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt (hunk 1 above)
Comment (kandersen) [reply compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/JvmLower.kt]: I pushed a revision, and think I have a design that avoids work in the normal case to the greatest extent possible.

**Comment by kandersen** [comment]
Hi @udalov ! Take a look at the redo of piping local declarations data around, and let me know if you are happy with the new design. I will squash the commits then.

**Comment by kandersen** [comment]
The corresponding plug-in side changes have been rebased on `kt-222-master` and can be seen here:
https://github.com/JetBrains/intellij-community/pull/1973

They are through review and are only pending this PR.

Let me know if there is anything else I can do to help this along!
